src/main/java/jp/co/soramitsu/sora/didresolver/domain/iroha/valueobjects/DID.java

```diff
@@ -0,0 +1,82 @@
+package jp.co.soramitsu.sora.didresolver.domain.iroha.valueobjects;
+
+import static jp.co.soramitsu.sora.didresolver.commons.DIDTypeEnum.ED;
+import static jp.co.soramitsu.sora.didresolver.commons.DIDTypeEnum.IROHA;
+import static jp.co.soramitsu.sora.didresolver.commons.DIDTypeEnum.UUID;
+
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import jp.co.soramitsu.sora.didresolver.commons.DIDTypeEnum;
+import jp.co.soramitsu.sora.didresolver.exceptions.DIDUnparseableException;
+import lombok.Value;
+
+@Value
+public class DID {
+
+  private static final String IROHA_REGEX = IROHA.getRegexp();
+  private static final String UUID_REGEX = UUID.getRegexp();
+  private static final String ED_REGEX = ED.getRegexp();
+
+  public static final Pattern didPattern = Pattern.compile(
+      "did:sora:(?<TYPE>iroha|uuid|ed):.+");
+  private static final Pattern irohaPattern = Pattern
+      .compile("did:sora:iroha:(?<IDENTIFIER>" + IROHA_REGEX + ")");
+  private static final Pattern uuidPattern = Pattern
+      .compile("did:sora:uuid:(?<IDENTIFIER>" + UUID_REGEX + ")");
+  private static final Pattern edPattern = Pattern
+      .compile("did:sora:ed:(?<IDENTIFIER>" + ED_REGEX + ")");
+
+
+  private DIDTypeEnum type;
+  private String identifier;
+  private String didString;
+
+  @Override
+  public String toString() {
+    return didString;
+  }
+
+  public DID(String did) throws DIDUnparseableException {
+    if (did != null) {
+
+      Matcher matcher = didPattern.matcher(did);
+      if (!matcher.matches()) {
+        throw new DIDUnparseableException(did);
+      }
+
+      String typeString = matcher.group("TYPE");
+
+      switch (typeString) {
```

Comment:
why don't you use dynamic creation of pattern? something like this pattern = Pattern.compile("did:sora:"+typeString+":(?<IDENTIFIER>" +DIDTypeEnum.valueOf(typeString.toUpperCase()).getRegexp() + ")");